airflow/bin/cli.py

```diff
@@ -336,6 +336,83 @@ def list_tasks(args, dag=None):
         print("\n".join(sorted(tasks)))
 
 
+def get_start_date(dag):
+    # Return dag start date if one exists
+    if dag.start_date:
+        return dag.start_date
+
+    # Otherwise. return the min task start date, which
+    # may come from the dag.default_args
+    task_start_dates = [t.start_date for t in dag.tasks]
+    if task_start_dates:
+        min_task_start_date = min(task_start_dates)
+    start_date = min_task_start_date if \
+        min_task_start_date else None
+    return start_date
+
+def get_end_date(dag):
+    # Return dag end date if one exists
+    if dag.end_date:
+        return dag.end_date
+
+    # Otherwise. return the dag.default_args
+    end_date = dag.default_args['end_date'] if \
+        (dag.default_args and 'end_date' in dag.default_args) else None
+    return end_date
+
+
+def parsedate_enhanced(args, dag):
+    '''
+    In addition to supporting date parsing, this also supports
+    @now,
+    @start_date (i.e. dag start date),
+    @end_date (i.e. dag end date if one exists), and
+    @latest (i.e. run closest to and before now that honors the schedule interval)
+    '''
+    timestr = args.execution_date
```

Comment:
I think execution_date is the only thing you use out of args, so maybe the function should just receive that. The function header should probably define a more generic `dttm_str` or something like that. Since it requires the DAG's context maybe it should be a DAG method.